### PR TITLE
Parallel contig assembly

### PR DESCRIPTION
**Pull request by woodymit**

### CODE REVIEW

File: genome_designer/genome_finish/tests/test_parallel_assembly.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+TEST_USERNAME = 'testuser'
+TEST_PASSWORD = 'password'
+TEST_EMAIL = 'test@example.com'
+
+GF_TEST_DIR = os.path.join(
+        settings.PWD,
+        'test_data/genome_finish_test')
+
+
+class TestParallelAssembly(TestCase):
```
Comment (glebkuznetsov) [inline genome_designer/genome_finish/tests/test_parallel_assembly.py]: FYI - Django tests use celery in synchronous mode. So this test is good for making sure semantics work, but it won't catch concurrency issues.

I tried setting up integration tests in the pasts (https://github.com/churchlab/millstone/blob/master/genome_designer/tests/integration/test_pipeline_integration.py#L31). These would bring up a celery server and run things in parallel. However, it caused way too many headaches and cryptic problems so we abandoned that route.

**Comment by glebkuznetsov** [comment]
Looks good. Test pass.

Brian, can you confirm that you've manually tested it with multiple cores in a case where there might be concurrency issues?

**Comment by woodymit** [comment]
Now I can confirm it passes manual testing.

**Comment by woodymit** [comment]
You can still break it though if you attempt to re-run an assembly that is running at the right moment.

File: genome_designer/genome_finish/insertion_placement_read_trkg.py [hunk 2]
```diff
@@ -472,11 +474,20 @@ def write_read_query_alignments_to_fastq(reads, fastq_path,
 
 def simple_align_with_bwa_mem(reads_fq, reference_fasta, output_bam_path):
 
+    if not os.path.exists(settings.TEMP_FILE_ROOT):
+            os.mkdir(settings.TEMP_FILE_ROOT)
```
Comment (glebkuznetsov) [inline genome_designer/genome_finish/insertion_placement_read_trkg.py]: Fix indentation.

File: genome_designer/genome_finish/insertion_placement_read_trkg.py [hunk 3]
```diff
@@ -472,11 +474,20 @@ def write_read_query_alignments_to_fastq(reads, fastq_path,
...
+
+    filename_prefix = 'temp_ref_copy'
```
Comment (glebkuznetsov) [inline genome_designer/genome_finish/insertion_placement_read_trkg.py]: What exactly is the bug? Ideally we would not generate copies of the reference genome.

At the least I would expect you to have a randomly generated name here (uuid). And then somewhere this temp reference should be deleted.

File: genome_designer/genome_finish/insertion_placement_read_trkg.py [hunk 4]
```diff
@@ -473,12 +474,7 @@ def write_read_query_alignments_to_fastq(reads, fastq_path,
 def simple_align_with_bwa_mem(reads_fq, reference_fasta, output_bam_path):
 
     # Ensure reference fasta is indexed
-    subprocess.call(' '.join([
-            '%s/bwa/bwa' % settings.TOOLS_DIR,
-            'index',
-            reference_fasta
-            ]),
-            shell=True, executable=settings.BASH_PATH)
+    ensure_bwa_index(reference_fasta)
```
Comment (glebkuznetsov) [inline genome_designer/genome_finish/insertion_placement_read_trkg.py]: This function might actually try to index and cause the threading issue we were debugging. Instead we should just have a hard assert here that checks for the existence of the index. Maybe make a new function in read_alignment_util.py called something like `has_bwa_index()` that returns a boolean which is the result of calling ` os.path.exists(ref_genome_fasta + '.bwt')`. And here, just `assert` on that result. If that fails, we have another bug somewhere above but we'll catch that immediately.

**Pull request merged**